File: backend/app/schemas/recognition.py
"""识别讲解相关 Pydantic Schema"""

from datetime import datetime
from pydantic import BaseModel
from app.schemas.common import PaginatedResponse


class CategoryCandidate(BaseModel):
    category: str
    confidence: float


class Explanation(BaseModel):
    history: str
    technique: str
    inheritor: str
    meaning: str


class CreationLink(BaseModel):
    style: str
    label: str


class ExhibitLink(BaseModel):
    id: int
    name: str


class HeatmapFeature(BaseModel):
    name: str
    x: float  # 归一化 x 坐标 (0-1)
    y: float  # 归一化 y 坐标 (0-1)
    label: str = ""


class RelatedRecommendations(BaseModel):
    creations: list[CreationLink]
    exhibits: list[ExhibitLink]


class RecognitionResponse(BaseModel):
    id: int
    image_url: str
    category: str
    confidence: float
    top3: list[CategoryCandidate]
    features: list[str]
    explanation: Explanation
    heatmap_url: str | None = None
    heatmap_data: list[HeatmapFeature] = []
    voice_url: str | None = None
    pattern_names: list[str] = []
    related: RelatedRecommendations
    created_at: datetime

    model_config = {"from_attributes": True}


class RecognitionListItem(BaseModel):
    id: int
    image_url: str
    category: str
    confidence: float
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/services/recognition.ts
import api from './api'

export interface RecognitionResult {
  id: number
  image_url: string
  category: string
  confidence: number
  top3: { category: string; confidence: number }[]
  features: string[]
  pattern_names: string[]
  explanation: {
    history: string
    technique: string
    inheritor: string
    meaning: string
  }
  heatmap_url: string | null
  heatmap_data: { name: string; x: number; y: number; label: string }[]
  voice_url: string | null
  related: {
    creations: { style: string; label: string }[]
    exhibits: { id: number; name: string }[]
  }
  created_at: string
}

export interface RecognitionListItem {
  id: number
  image_url: string
  category: string
  confidence: number
  created_at: string
}

export async function uploadAndRecognize(
  file: File,
  onProgress?: (percent: number) => void,
): Promise<RecognitionResult> {
  const formData = new FormData()
  formData.append('file', file)
  const res = await api.post('/api/recognition/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 120000,
    onUploadProgress: (progressEvent) => {
      if (onProgress && progressEvent.total) {
        const percent = Math.round((progressEvent.loaded * 100) / progressEvent.total)
        onProgress(percent)
      }
    },
  })
  return res.data
}

export async function getHistory(page = 1, pageSize = 10) {
  const res = await api.get('/api/recognition/history', {
    params: { page, page_size: pageSize },
  })
  return res.data
}

export async function getDetail(id: number): Promise<RecognitionResult> {
  const res = await api.get(`/api/recognition/${id}`)
  return res.data
}
